pkgs/development/libraries/glibc/ld-fallback.patch

```diff
@@ -0,0 +1,275 @@
+
+Thus, LD_FALLBACK_PATH provides a way to specify fallback library paths.
+
+Consider the following use cases:
+
+- Running a closed-source program (like Steam and its games) on an unsupported
+  distribution: closed-source programs commonly override LD_LIBRARY_PATH to
+  override system libraries with their own bundled versions, but they
+  frequently do it the wrong way by overriding it while ignoring the current
+  value of LD_LIBRARY_PATH, thus failing to pick up any non-system libraries
+  they don't bundle but require specific versions of.
+
+  LD_FALLBACK_PATH, being the low-priority option, can't be used to achieve
+  the same effect as LD_LIBRARY_PATH from the perspective of the closed-source
+  program's vendor, thus making it unlikely they would ever touch
+  LD_FALLBACK_PATH. But it can be used by other tools to supply needed default
+  libraries to the closed-source program in question.
+
+- Similarly, sometimes you just want to change the underlying library for a
+  set of programs without changing the system default. For instance, if you
+  have different GPUs requiring different libGL's (though, specifically for
+  libGL there are several less ad-hoc solutions than this).
+
+In general, before LD_FALLBACK_PATH there was no good way to solve these
+problems without generating a whole chroot. Now you can collect those
+libraries you want to override system defaults but not override those found
+via LD_LIBRARY_PATH in a directory somewhere and put its path into
+LD_FALLBACK_PATH.
+---
+ elf/dl-load.c               | 60 +++++++++++++++++++++++++++++++++++--
+ elf/dl-main.h               |  7 +++++
+ elf/dl-support.c            |  1 +
+ elf/dl-usage.c              |  2 ++
+ elf/rtld.c                  | 22 ++++++++++++++
+ sysdeps/generic/ldsodefs.h  |  1 +
+ sysdeps/generic/unsecvars.h |  1 +
+ 7 files changed, 92 insertions(+), 2 deletions(-)
+
+diff --git a/elf/dl-load.c b/elf/dl-load.c
+index 9a87fda9c9..a8fd639238 100644
+--- a/elf/dl-load.c
++++ b/elf/dl-load.c
+@@ -101,6 +101,8 @@ int __stack_prot attribute_hidden attribute_relro
+ 
+ /* This is the decomposed LD_LIBRARY_PATH search path.  */
+ struct r_search_path_struct __rtld_env_path_list attribute_relro;
++/* This is the decomposed LD_FALLBACK_PATH search path.  */
++struct r_search_path_struct __rtld_fallback_path_list attribute_relro;
+ 
+ /* List of the hardware capabilities we might end up using.  */
+ #ifdef SHARED
+@@ -701,6 +703,7 @@ cache_rpath (struct link_map *l,
+ 
+ void
+ _dl_init_paths (const char *llp, const char *source,
++		const char *lfp, const char *fsource,
+ 		const char *glibc_hwcaps_prepend,
+ 		const char *glibc_hwcaps_mask)
+ {
+@@ -854,6 +857,45 @@ _dl_init_paths (const char *llp, const char *source,
+     }
+   else
+     __rtld_env_path_list.dirs = (void *) -1;
++
++  /* Made by copying the above followed by
++      %s/llp/lfp/g
++      %s/env_path_list/fallback_path_list/g
++      %s/source/fsource/g
++   */
```

Comment:
Instead of duplicating this block, you might be able to insert (approximately) this text:

```
  scan_dl_path(llp, env_path_list, source);
  scan_dl_path(lfp, fallback_path_list, fsource);
}

void
scan_dl_path(char* llp, char** env_path_list, char* source)
{
```

... and get any future changes to the copy-pasted block "for free".

Reply:
Good idea!